sessiontxn/interface.go

```diff
@@ -15,71 +15,90 @@
+	TxnMode string
+	// causalConsistencyOnly means whether enable causal consistency for transactions, default is false
+	CausalConsistencyOnly bool
+	// staleReadTS indicates the read ts for the stale read transaction.
+	//The default value is zero which means not a stale read transaction.
+	StaleReadTS uint64
 }
 
-// GetReadTS returns the read timestamp used by select statement (not for select ... for update)
-func (p *SimpleTxnContextProvider) GetReadTS() (uint64, error) {
-	if p.GetReadTSFunc == nil {
-		return 0, errors.New("ReadTSFunc not set")
-	}
-	return p.GetReadTSFunc()
-}
+// TxnContextProvider provides txn context
+type TxnContextProvider interface {
+	// GetTxnInfoSchema returns the information schema used by txn
+	GetTxnInfoSchema() infoschema.InfoSchema
+	// GetStmtReadTS returns the read timestamp used by select statement (not for select ... for update)
+	GetStmtReadTS() (uint64, error)
+	// GetStmtForUpdateTS returns the read timestamp used by update/insert/delete or select ... for update
+	GetStmtForUpdateTS() (uint64, error)
 
-// GetForUpdateTS returns the read timestamp used by update/insert/delete or select ... for update
-func (p *SimpleTxnContextProvider) GetForUpdateTS() (uint64, error) {
-	if p.GetForUpdateTSFunc == nil {
-		return 0, errors.New("GetForUpdateTSFunc not set")
-	}
-	return p.GetForUpdateTSFunc()
+	// OnInitialize is the hook that should be called when enter a new txn with this provider
+	OnInitialize(ctx context.Context, enterNewTxnType EnterNewTxnType) error
+	// OnStmtStart is the hook that should be called when a new statement started
+	OnStmtStart(ctx context.Context) error
 }
 
 // TxnManager is an interface providing txn context management in session
 type TxnManager interface {
 	// GetTxnInfoSchema returns the information schema used by txn
 	GetTxnInfoSchema() infoschema.InfoSchema
-	// GetReadTS returns the read timestamp used by select statement (not for select ... for update)
-	GetReadTS() (uint64, error)
-	// GetForUpdateTS returns the read timestamp used by update/insert/delete or select ... for update
-	GetForUpdateTS() (uint64, error)
-
+	// GetStmtReadTS returns the read timestamp used by select statement (not for select ... for update)
+	GetStmtReadTS() (uint64, error)
+	// GetStmtForUpdateTS returns the read timestamp used by update/insert/delete or select ... for update
+	GetStmtForUpdateTS() (uint64, error)
 	// GetContextProvider returns the current TxnContextProvider
 	GetContextProvider() TxnContextProvider
-	// SetContextProvider sets the context provider
-	SetContextProvider(provider TxnContextProvider) error
+
+	// EnterNewTxn enters a new transaction.
+	EnterNewTxn(ctx context.Context, req *EnterNewTxnRequest) error
+	// OnStmtStart is the hook that should be called when a new statement started
+	OnStmtStart(ctx context.Context) error
+}
+
+// NewTxn starts a new optimistic and active txn
+func NewTxn(ctx context.Context, sctx sessionctx.Context) error {
```

Comment:
Do we need to add comments to remind users they should not use these two methods `NewTxn` and `NewTxnInStmt` directly.

Reply:
What do you mean by "should not use these two methods NewTxn and NewTxnInStmt directly"?

Reply:
For example, some background workers are needed for new features, and in which transactional interfaces will be used to do something, we need to add comments to explain when should or should not we use `NewTxn` here.

Reply:
done